mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java

```diff
@@ -0,0 +1,525 @@
+    private Schema expectedSchema;
+    private Schema tableSchema;
+    private InMemoryDataModel inMemoryDataModel;
+    private Closeable currentCloseable;
+    private boolean reuseContainers;
+    private boolean caseSensitive;
+
+    @Override
+    public void initialize(InputSplit split, TaskAttemptContext newContext) {
+      Configuration conf = newContext.getConfiguration();
+      // For now IcebergInputFormat does its own split planning and does not
+      // accept FileSplit instances
+      CombinedScanTask task = ((IcebergSplit) split).task;
+      this.context = newContext;
+      this.tasks = task.files().iterator();
+      this.tableSchema = SchemaParser.fromJson(conf.get(TABLE_SCHEMA));
+      String readSchemaStr = conf.get(READ_SCHEMA);
+      if (readSchemaStr != null) {
+        this.expectedSchema = SchemaParser.fromJson(readSchemaStr);
+      }
+      this.reuseContainers = conf.getBoolean(REUSE_CONTAINERS, false);
+      this.caseSensitive = conf.getBoolean(CASE_SENSITIVE, true);
+      this.inMemoryDataModel = conf.getEnum(IN_MEMORY_DATA_MODEL, InMemoryDataModel.DEFAULT);
+      this.currentIterator = open(tasks.next());
+    }
+
+    @Override
+    public boolean nextKeyValue() throws IOException {
+      while (true) {
+        if (currentIterator.hasNext()) {
+          currentRow = currentIterator.next();
+          return true;
+        } else if (tasks.hasNext()) {
+          currentCloseable.close();
+          currentIterator = open(tasks.next());
+        } else {
+          return false;
+        }
+      }
+    }
+
+    @Override
+    public Void getCurrentKey() {
+      return null;
+    }
+
+    @Override
+    public T getCurrentValue() {
+      return currentRow;
+    }
+
+    @Override
+    public float getProgress() {
+      return context.getProgress();
+    }
+
+    @Override
+    public void close() throws IOException {
+      currentCloseable.close();
+    }
+
+    private Iterator<T> open(FileScanTask currentTask) {
+      DataFile file = currentTask.file();
+      // schema of rows returned by readers
+      PartitionSpec spec = currentTask.spec();
+      Set<Integer> idColumns = spec.identitySourceIds();
+      Schema readSchema = expectedSchema != null ? expectedSchema : tableSchema;
+      boolean hasJoinedPartitionColumns = !idColumns.isEmpty();
+      if (hasJoinedPartitionColumns) {
+        readSchema = TypeUtil.selectNot(tableSchema, idColumns);
+        Schema identityPartitionSchema = TypeUtil.select(tableSchema, idColumns);
+        return Iterators.transform(
+            open(currentTask, readSchema),
+            row -> withPartitionColumns(row, identityPartitionSchema, spec, file.partition()));
+      } else {
+        return open(currentTask, readSchema);
+      }
+    }
+
+    private Iterator<T> open(FileScanTask currentTask, Schema readSchema) {
+      DataFile file = currentTask.file();
+      // TODO should we somehow make use of FileIO to create inputFile?
```

Comment:
Yes. `FileIO` is `Serializable` and should be used to create input files. We should also use the encryption manager to be correct (could open a follow-up).

Reply:
I'll create a follow up for this if that's OK. Should this be serialized as base64 and added to job conf. I think job configuration generally has a default size limit of 5MB

Reply:
Yeah, sounds good.

Reply:
Created a follow up #865